File: app/src/main/java/com/example/polyquicktrade/ui/AddProductViewModel.java
package com.example.polyquicktrade.ui;

import android.net.Uri;
import android.util.Log;

import com.example.polyquicktrade.pojo.Enquiry;
import com.example.polyquicktrade.pojo.Product;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import java.io.File;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

public class AddProductViewModel extends ViewModel {

    public MutableLiveData<Product> getmProduct() {
        return mProduct;
    }

    private MutableLiveData<Product> mProduct;

    public AddProductViewModel(){
        mProduct = new MutableLiveData<>();


    }

    public void uploadProductToFirestone(final Product product, FirebaseFirestore db) {


        Date date = new Date(product.getUploadTime());
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);


        int month = cal.get(Calendar.MONTH);
        int week = cal.get(Calendar.WEEK_OF_MONTH);
        int day = cal.get(Calendar.DAY_OF_WEEK);


        //Database storage: Products C-> Category D-> Month C-> Week D-> Day C-> Product D
        DocumentReference newProdRef = db
                .collection("Products")
                .document(product.getCategory())
                .collection("Month_" + month)
                .document("Week_" + week)
                .collection("product_documents")
                .document();

        product.setId(newProdRef.getId());

        newProdRef
                .set(product)
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        Log.d("MY_TAG", "Product Succesfully Written!");
                        AddProductViewModel.this.mProduct.setValue(product);



                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.w("MY_TAG", "Error adding Product", e);

                    }
                });

    }
    public void uploadProductToFirestone(final Product product, final FirebaseFirestore db, final String enquiryID) {


        Date date = new Date(product.getUploadTime());
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);


        int month = cal.get(Calendar.MONTH);
        int week = cal.get(Calendar.WEEK_OF_MONTH);
        int day = cal.get(Calendar.DAY_OF_WEEK);


        //Database storage: Products C-> Category D-> Month C-> Week D-> Day C-> Product D
        DocumentReference newProdRef = db
                .collection("Products")
                .document(product.getCategory())
                .collection("Month_" + month)
                .document("Week_" + week)
                .collection("product_documents")
                .document();

        product.setId(newProdRef.getId());

        newProdRef
                .set(product)
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        Log.d("MY_TAG", "Product Succesfully Written!");
                        AddProductViewModel.this.mProduct.setValue(product);

                        addProductIDtoEnquiry(enquiryID, product.getId(), db);



                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.w("MY_TAG", "Error adding Product", e);

                    }
                });

    }

    private void addProductIDtoEnquiry(String enquiryID, final String productID, final FirebaseFirestore db) {

        final ArrayList<Enquiry> enquiries = new ArrayList<>();
        db.collectionGroup("enquiry_documents").whereEqualTo("id", enquiryID).get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {

                        if (task.isSuccessful()){

                            for (QueryDocumentSnapshot document : Objects.requireNonNull(task.getResult())) {
                                Log.d("MY_TAG", document.getId() + " => " + document.getData());
                                enquiries.add(document.toObject(Enquiry.class));

                            }
                            updateEnquiryProducts(enquiries.get(0), productID, db);

                        }
                        else {
                            Log.d("MY_TAG", "AskViewModel task not successful: ");

                        }
                    }
                }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {

                Log.d("MY_TAG", "Message: "+e.getMessage());
                Log.d("MY_TAG", "eString: "+ e.toString());

            }
        });
    }

    private void updateEnquiryProducts(Enquiry enquiry, String productID, FirebaseFirestore db) {


            Date date = new Date(enquiry.getCreationTime());
            Calendar cal = Calendar.getInstance();
            cal.setTime(date);


            int month = cal.get(Calendar.MONTH);


        //Database Storage: Enquiries C-> Month D-> Week C -> Enquiry D
            DocumentReference enqRef = db
                    .collection("Enquiry")
                    .document("Month_" + month)
                    .collection("enquiry_documents")
                    .document(enquiry.getId());


            if (            enquiry.getProductResponses() !=null){
                ArrayList<String> productIDs = enquiry.getProductResponses();
                productIDs.add(productID);
                enquiry.setProductResponses(productIDs);

            }else {
                ArrayList<String> productIDs = new ArrayList<>();
                productIDs.add(productID);
                enquiry.setProductResponses(productIDs);
            }

            enqRef.set(enquiry).addOnSuccessListener(new OnSuccessListener<Void>() {
                @Override
                public void onSuccess(Void aVoid) {
                   Log.d("MY_TAG", "Enquiry Updated");

                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {

                }
            });






    }

    public void uploadToFirebase(StorageReference storageRef, final Product product, final FirebaseFirestore db){

        Date date = new Date(product.getUploadTime());
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);


        int month = cal.get(Calendar.MONTH);
        int week = cal.get(Calendar.WEEK_OF_MONTH);
        int day = cal.get(Calendar.DAY_OF_WEEK);

        Uri file = Uri.fromFile(new File(product.getPhotoURI()));


        final StorageReference imageRef = storageRef.
                child("Product").
                child(product.getCategory()).
                child(month + "")
                .child(week + "")
                .child(day + "")
                .child(Objects.requireNonNull(file.getLastPathSegment()));

        UploadTask uploadTask = imageRef.putFile(file);
//
//        Task<Uri> urlTask = uploadTask.continueWithTask(new Continuation<UploadTask.TaskSnapshot, Task<Uri>>() {
//            @Override
//            public Task<Uri> then(@NonNull Task<UploadTask.TaskSnapshot> task) throws Exception {
//                if (!task.isSuccessful()) {
//                    throw task.getException();
//                }
//
//                // Continue with the task to get the download URL
//                return imageRef.getDownloadUrl();
//            }
//        }).addOnCompleteListener(new OnCompleteListener<Uri>() {
//            @Override
//            public void onComplete(@NonNull Task<Uri> task) {
//                if (task.isSuccessful()) {
//                    Uri downloadUri = task.getResult();
//                } else {
//                    // Handle failures
//                    // ...
//                }
//            }
//        });

//        imageRef.getDownloadUrl().addOnCompleteListener(new OnCompleteListener<Uri>() {
//            @Override
//            public void onComplete(@NonNull Task<Uri> task) {
//                product.setPhotoURI(
//                        task.getResult().toString());
//
//
//                uploadProductToFirestone(product, db);
//                Log.w("MY_TAG", "succesfully added image to cloud storage");
//                Log.w("MY_TAG", task.getResult();
//            }
//        }).addOnFailureListener(new OnFailureListener() {
//            @Override
//            public void onFailure(@NonNull Exception e) {
//                Log.w("MY_TAG", "Error adding image to cloud storage", e);
//
//            }
//        });

        uploadTask.addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
            @Override
            public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {

                taskSnapshot.getStorage().getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                    @Override
                    public void onSuccess(Uri uri) {
                        product.setPhotoURI(uri.toString());


                        uploadProductToFirestone(product, db);
                        Log.w("MY_TAG", "succesfully added image to cloud storage");
                        Log.w("MY_TAG", uri.toString());
                    }
                });


            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {

                Log.w("MY_TAG", "Error adding image to cloud storage", e);
            }
        });



    }
    public void uploadToFirebase(StorageReference storageRef, final Product product, final FirebaseFirestore db, final String enquiryID){

        Date date = new Date(product.getUploadTime());
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);


        int month = cal.get(Calendar.MONTH);
        int week = cal.get(Calendar.WEEK_OF_MONTH);
        int day = cal.get(Calendar.DAY_OF_WEEK);

        Uri file = Uri.fromFile(new File(product.getPhotoURI()));


        final StorageReference imageRef = storageRef.
                child("Product").
                child(product.getCategory()).
                child(month + "")
                .child(week + "")
                .child(day + "")
                .child(file.getLastPathSegment());

        UploadTask uploadTask = imageRef.putFile(file);


        uploadTask.addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
            @Override
            public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {

                taskSnapshot.getStorage().getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                    @Override
                    public void onSuccess(Uri uri) {
                        product.setPhotoURI(uri.toString());


                        uploadProductToFirestone(product, db, enquiryID);
                        Log.w("MY_TAG", "succesfully added image to cloud storage");
                        Log.w("MY_TAG", uri.toString());
                    }
                });


            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {

                Log.w("MY_TAG", "Error adding image to cloud storage", e);
            }
        });



    }


}

File: app/src/main/java/com/example/polyquicktrade/ui/AddProductViewModel.kt
package com.example.polyquicktrade.ui

import android.net.Uri
import android.util.Log
import com.example.polyquicktrade.pojo.Product.uploadTime
import com.example.polyquicktrade.pojo.Product.category
import com.example.polyquicktrade.pojo.Product.id
import com.example.polyquicktrade.pojo.Enquiry.creationTime
import com.example.polyquicktrade.pojo.Enquiry.id
import com.example.polyquicktrade.pojo.Enquiry.productResponses
import com.example.polyquicktrade.pojo.Product.photoURI
import androidx.lifecycle.ViewModel
import androidx.lifecycle.MutableLiveData
import com.example.polyquicktrade.pojo.Product
import com.google.firebase.firestore.FirebaseFirestore
import com.google.android.gms.tasks.OnSuccessListener
import com.google.android.gms.tasks.OnFailureListener
import com.example.polyquicktrade.pojo.Enquiry
import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.firestore.QuerySnapshot
import com.google.firebase.firestore.QueryDocumentSnapshot
import com.google.firebase.storage.StorageReference
import com.google.firebase.storage.UploadTask
import java.io.File
import java.util.*

class AddProductViewModel : ViewModel() {
    fun getmProduct(): MutableLiveData<Product> {
        return mProduct
    }

    private val mProduct: MutableLiveData<Product>
    fun uploadProductToFirestone(product: Product, db: FirebaseFirestore) {
        val date = Date(product.uploadTime)
        val cal = Calendar.getInstance()
        cal.time = date
        val month = cal[Calendar.MONTH]
        val week = cal[Calendar.WEEK_OF_MONTH]
        val day = cal[Calendar.DAY_OF_WEEK]


        //Database storage: Products C-> Category D-> Month C-> Week D-> Day C-> Product D
        val newProdRef = db
            .collection("Products")
            .document(product.category)
            .collection("Month_$month")
            .document("Week_$week")
            .collection("product_documents")
            .document()
        product.id = newProdRef.id
        newProdRef
            .set(product)
            .addOnSuccessListener {
                Log.d("MY_TAG", "Product Succesfully Written!")
                mProduct.value = product
            }
            .addOnFailureListener { e -> Log.w("MY_TAG", "Error adding Product", e) }
    }

    fun uploadProductToFirestone(product: Product, db: FirebaseFirestore, enquiryID: String) {
        val date = Date(product.uploadTime)
        val cal = Calendar.getInstance()
        cal.time = date
        val month = cal[Calendar.MONTH]
        val week = cal[Calendar.WEEK_OF_MONTH]
        val day = cal[Calendar.DAY_OF_WEEK]


        //Database storage: Products C-> Category D-> Month C-> Week D-> Day C-> Product D
        val newProdRef = db
            .collection("Products")
            .document(product.category)
            .collection("Month_$month")
            .document("Week_$week")
            .collection("product_documents")
            .document()
        product.id = newProdRef.id
        newProdRef
            .set(product)
            .addOnSuccessListener {
                Log.d("MY_TAG", "Product Succesfully Written!")
                mProduct.value = product
                addProductIDtoEnquiry(enquiryID, product.id, db)
            }
            .addOnFailureListener { e -> Log.w("MY_TAG", "Error adding Product", e) }
    }

    private fun addProductIDtoEnquiry(enquiryID: String, productID: String, db: FirebaseFirestore) {
        val enquiries = ArrayList<Enquiry>()
        db.collectionGroup("enquiry_documents").whereEqualTo("id", enquiryID).get()
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    for (document in Objects.requireNonNull(task.result)) {
                        Log.d("MY_TAG", document.id + " => " + document.data)
                        enquiries.add(document.toObject(Enquiry::class.java))
                    }
                    updateEnquiryProducts(enquiries[0], productID, db)
                } else {
                    Log.d("MY_TAG", "AskViewModel task not successful: ")
                }
            }.addOnFailureListener { e ->
                Log.d("MY_TAG", "Message: " + e.message)
                Log.d("MY_TAG", "eString: $e")
            }
    }

    private fun updateEnquiryProducts(enquiry: Enquiry, productID: String, db: FirebaseFirestore) {
        val date = Date(enquiry.creationTime)
        val cal = Calendar.getInstance()
        cal.time = date
        val month = cal[Calendar.MONTH]


        //Database Storage: Enquiries C-> Month D-> Week C -> Enquiry D
        val enqRef = db
            .collection("Enquiry")
            .document("Month_$month")
            .collection("enquiry_documents")
            .document(enquiry.id!!)
        if (enquiry.productResponses != null) {
            val productIDs = enquiry.productResponses
            productIDs!!.add(productID)
            enquiry.productResponses = productIDs
        } else {
            val productIDs = ArrayList<String>()
            productIDs.add(productID)
            enquiry.productResponses = productIDs
        }
        enqRef.set(enquiry).addOnSuccessListener { Log.d("MY_TAG", "Enquiry Updated") }
            .addOnFailureListener { }
    }

    fun uploadToFirebase(storageRef: StorageReference, product: Product, db: FirebaseFirestore) {
        val date = Date(product.uploadTime)
        val cal = Calendar.getInstance()
        cal.time = date
        val month = cal[Calendar.MONTH]
        val week = cal[Calendar.WEEK_OF_MONTH]
        val day = cal[Calendar.DAY_OF_WEEK]
        val file = Uri.fromFile(File(product.photoURI))
        val imageRef =
            storageRef.child("Product").child(product.category).child(month.toString() + "")
                .child(week.toString() + "")
                .child(day.toString() + "")
                .child(file.lastPathSegment!!)
        val uploadTask = imageRef.putFile(file)
        //
//        Task<Uri> urlTask = uploadTask.continueWithTask(new Continuation<UploadTask.TaskSnapshot, Task<Uri>>() {
//            @Override
//            public Task<Uri> then(@NonNull Task<UploadTask.TaskSnapshot> task) throws Exception {
//                if (!task.isSuccessful()) {
//                    throw task.getException();
//                }
//
//                // Continue with the task to get the download URL
//                return imageRef.getDownloadUrl();
//            }
//        }).addOnCompleteListener(new OnCompleteListener<Uri>() {
//            @Override
//            public void onComplete(@NonNull Task<Uri> task) {
//                if (task.isSuccessful()) {
//                    Uri downloadUri = task.getResult();
//                } else {
//                    // Handle failures
//                    // ...
//                }
//            }
//        });

//        imageRef.getDownloadUrl().addOnCompleteListener(new OnCompleteListener<Uri>() {
//            @Override
//            public void onComplete(@NonNull Task<Uri> task) {
//                product.setPhotoURI(
//                        task.getResult().toString());
//
//
//                uploadProductToFirestone(product, db);
//                Log.w("MY_TAG", "succesfully added image to cloud storage");
//                Log.w("MY_TAG", task.getResult();
//            }
//        }).addOnFailureListener(new OnFailureListener() {
//            @Override
//            public void onFailure(@NonNull Exception e) {
//                Log.w("MY_TAG", "Error adding image to cloud storage", e);
//
//            }
//        });
        uploadTask.addOnSuccessListener { taskSnapshot ->
            taskSnapshot.storage.downloadUrl.addOnSuccessListener { uri ->
                product.photoURI = uri.toString()
                uploadProductToFirestone(product, db)
                Log.w("MY_TAG", "succesfully added image to cloud storage")
                Log.w("MY_TAG", uri.toString())
            }
        }.addOnFailureListener { e -> Log.w("MY_TAG", "Error adding image to cloud storage", e) }
    }

    fun uploadToFirebase(
        storageRef: StorageReference,
        product: Product,
        db: FirebaseFirestore,
        enquiryID: String
    ) {
        val date = Date(product.uploadTime)
        val cal = Calendar.getInstance()
        cal.time = date
        val month = cal[Calendar.MONTH]
        val week = cal[Calendar.WEEK_OF_MONTH]
        val day = cal[Calendar.DAY_OF_WEEK]
        val file = Uri.fromFile(File(product.photoURI))
        val imageRef =
            storageRef.child("Product").child(product.category).child(month.toString() + "")
                .child(week.toString() + "")
                .child(day.toString() + "")
                .child(file.lastPathSegment!!)
        val uploadTask = imageRef.putFile(file)
        uploadTask.addOnSuccessListener { taskSnapshot ->
            taskSnapshot.storage.downloadUrl.addOnSuccessListener { uri ->
                product.photoURI = uri.toString()
                uploadProductToFirestone(product, db, enquiryID)
                Log.w("MY_TAG", "succesfully added image to cloud storage")
                Log.w("MY_TAG", uri.toString())
            }
        }.addOnFailureListener { e -> Log.w("MY_TAG", "Error adding image to cloud storage", e) }
    }

    init {
        mProduct = MutableLiveData()
    }
}